### ISSUE TITLE
Rust 1.18 ICE regression: assertion failed: !unsize (nightly regression #41604)

### ISSUE THREAD

**Issue by xfix** [rust-lang/rust#41944]
When compiling image with beta, the following error happens.

```
error: internal compiler error: unexpected panic
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports
thread 'rustc' panicked at 'assertion failed: !unsize', /checkout/src/librustc_typeck/check/mod.rs:2185
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: Could not compile `image`.
```

This was reported in nightly as #41604 and fixed in nightly. Reported by https://users.rust-lang.org/t/how-to-report-problems-with-rust-beta/10759?u=xfix.

**Comment by johnthagen** [comment]
We just hit this as well.

Travis CI full example: https://travis-ci.org/johnthagen/rust-belt/jobs/232185624
AppVeyor full example: https://ci.appveyor.com/project/johnthagen/rust-belt/build/1.0.486/job/k7ao14uh4gh032y5

**Comment by Mark-Simulacrum** [comment]
I'm confused. As far as I can tell, this is a known problem -- https://github.com/rust-lang/rust/issues/41604 -- and is awaiting a backport to beta, which would then fix it there. Shouldn't this be closed?

**Comment by johnthagen** [comment]
@Mark-Simulacrum I'm totally fine with closing if it's a known, solved issue. I just hadn't seen that fix after looking around and wanting to err on the side of reporting.

**Comment by Mark-Simulacrum** [comment]
Oh, yeah, please do report, it's better to have the report than not have it and miss a regression. However, since I know a lot of people hit this (both on nightly and beta) I'll keep this open to allow them to land here. We can close it in the potential future beta backport PR.

**Comment by michaelwoerister** [comment]
Once https://github.com/rust-lang/rust/pull/42053 lands, this should also be fixed on beta.

**Comment by brson** [comment]
Also affects life-0.1.2 cc @lj-ditrapani

**Comment by nikomatsakis** [comment]
I think I have verified that life-0.1.2 is fixed by https://github.com/rust-lang/rust/pull/42053. Though I wonder if I've messed up my testing setup.

**Comment by brson** [comment]
Thanks @nikomatsakis. I've also confirmed that life is fixed on beta.
